src/Features/Core/Portable/Completion/CommonCompletionItem.cs

```diff
@@ -0,0 +1,123 @@
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal static class CommonCompletionItem 
+    {
+        public static CompletionItem Create(
+            string displayText,
+            TextSpan span,
+            Glyph? glyph = null,
+            ImmutableArray<SymbolDisplayPart> description = default(ImmutableArray<SymbolDisplayPart>),
+            string sortText = null,
+            string filterText = null,
+            bool preselect = false,
+            bool showsWarningIcon = false,
+            bool shouldFormatOnCommit = false,
+            bool isArgumentName = false,
+            ImmutableDictionary<string, string> properties = null,
+            ImmutableArray<string> tags = default(ImmutableArray<string>),
+            CompletionItemRules rules = null)
+        {
+            tags = tags.IsDefault ? ImmutableArray<string>.Empty : tags;
+
+            if (glyph != null)
+            {
+                // put glyph tags first
+                tags = GlyphTags.GetTags(glyph.Value).AddRange(tags);
+            }
+
+            if (showsWarningIcon)
+            {
+                tags = tags.Add(CompletionTags.Warning);
+            }
+
+            if (isArgumentName)
+            {
+                tags = tags.Add(CompletionTags.ArgumentName);
+            }
+
+            properties = properties ?? ImmutableDictionary<string, string>.Empty;
+            if (!description.IsDefault && description.Length > 0)
+            {
+                properties = properties.Add("Description", EncodeDescription(description));
+            }
+
+            rules = rules ?? CompletionItemRules.Default;
+            rules = rules.WithPreselect(preselect)
+                         .WithFormatOnCommit(shouldFormatOnCommit);
+
+            return CompletionItem.Create(
+                displayText: displayText,
+                filterText: filterText,
+                sortText: sortText,
+                span: span,
+                properties: properties,
+                tags: tags,
+                rules: rules);
+        }
+
+        public static bool HasDescription(CompletionItem item)
+        {
+            return item.Properties.ContainsKey("Description");
+        }
+
+        public static CompletionDescription GetDescription(CompletionItem item)
+        {
+            string encodedDescription;
+            if (item.Properties.TryGetValue("Description", out encodedDescription))
+            {
+                return DecodeDescription(encodedDescription);
+            }
+            else
+            {
+                return CompletionDescription.Empty;
+            }
+        }
+
+        private static char[] s_descriptionSeparators = new char[] { '|' };
```

Comment:
How does this work if the description actually contains a | in it?  Why not use something like JSON so that we can roundtrip anything?